### PR TITLE
【Infer Symbolic Shape BUAA No.10 No.73 No.78】Add mode , margin_cross_entropy , box_clip. op 

### PR DESCRIPTION
**Pull request by crazyxiaoxi**
### PR Category
CINN

### PR Types
improvements

### Description
添加mode , margin_cross_entropy,box_clip算子符号推导接口
均已检查Optest

### CODE REVIEW

**Comment by crazyxiaoxi** [comment]
test_index_select_strided.py 中已包含OpTest单测，self.check_output() 确认使用，申请approve流水线

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 1]
```diff
@@ -563,6 +563,29 @@ bool MatmulOpInferSymbolicShape(pir::Operation *op,
...
+bool MarginCrossEntropyOpInferSymbolicShape(
+    pir::Operation *op, pir::InferSymbolicShapeContext *infer_context) {
+  const auto &logits_shape_or_data =
+      infer_context->GetShapeOrDataForValue(op->operand_source(0));
+
+  std::vector<symbol::DimExpr> logits_dims = logits_shape_or_data.shape();
+
+  size_t logits_rank = logits_dims.size();
+  auto axis = logits_rank - 1;
+
```
Comment (gongshaotian) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc:790]: 这里参考infermeta 加一下 equal constrain 约束

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 2]
```diff
@@ -720,6 +720,46 @@ bool MatmulOpInferSymbolicShape(pir::Operation *op,
...
+  std::vector<symbol::DimExpr> logits_dims = logits_shape_or_data.shape();
+  std::vector<symbol::DimExpr> labels_dims = labels_shape_or_data.shape();
+
+  size_t logits_rank = logits_dims.size();
+  auto axis = logits_rank - 1;
+
+  for (size_t i = 0; i < logits_rank; i++) {
+    if (i != axis) {
+      infer_context->AddBroadcastableCstr(logits_dims[i], labels_dims[i]);
```
Comment (zyfncg) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 这里应该是 Equal 约束吧？

**Review by Hongqing-work — APPROVED** [review]

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 3]
```diff
@@ -720,6 +720,46 @@ bool MatmulOpInferSymbolicShape(pir::Operation *op,
...
+    }
+  }
+
+  const auto one = symbol::DimExpr{1};
+
+  if (labels_dims.size() > 1) {
+    infer_context->AddEqualCstr(logits_dims[axis], one);
+  }
```
Comment (Hongqing-work) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 这里逻辑和infer meta不太符合，和之前的相等约束一并修改后再合并吧

**Review by Hongqing-work — CHANGES_REQUESTED** [review]

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc (hunk 3 above)
Comment (crazyxiaoxi) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 已修改

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc (hunk 2 above)
Comment (crazyxiaoxi) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc]: 已修改

**Comment by crazyxiaoxi** [comment]
把mode和boxclip放在一起测试，
原pr mode见https://github.com/PaddlePaddle/Paddle/pull/66621
原pr boxclip见https://github.com/PaddlePaddle/Paddle/pull/66629
test_mode_op.py中已包含OpTest单测，self.check_output() 确认使用，申请approve流水线
截屏2024-08-09 10 42 50
test_box_clip_op.py中已包含OpTest单测，本地测试通过
截屏2024-08-09 11 10 10

**Review by gongshaotian — APPROVED** [review]
LGTM

File: paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc [hunk 4]
```diff
@@ -735,14 +735,14 @@ bool MarginCrossEntropyOpInferSymbolicShape(
...
+      infer_context->AddEqualCstr(logits_dims[i], labels_dims[i]);
     }
   }
 
-  const auto one = symbol::DimExpr{1};
+  const auto &one = symbol::DimExpr{1};
 
   if (labels_dims.size() > 1) {
-    infer_context->AddEqualCstr(logits_dims[axis], one);
+    infer_context->AddEqualCstr(labels_dims[axis - 1], one);
```
Comment (Hongqing-work) [inline paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc:800]: infer meta里是检查labels_dims[logits_rank - 1]是否等于1，再确认下是否一致吧

Comment (crazyxiaoxi) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc:800]: 上面是infermeta的定义
截屏2024-08-12 11 39 34
检查labels_dims[logits_rank - 1]是否等于1


下面是opinfersymbolicshape
截屏2024-08-12 11 35 36

确实在写的时候不需要 ' axis-1' 只用axis = logits_rank -1

Comment (Hongqing-work) [reply paddle/fluid/pir/dialect/operator/interface/infer_symbolic_shape/binary_infer_sym.cc:800]: 那下一个PR顺带修复一下吧，提的时候把这个PR链接带上

**Review by Hongqing-work — APPROVED** [review]
